components/normalizer/src/lib.rs

```diff
@@ -1476,6 +1483,13 @@ macro_rules! normalizer_methods {
             ret
         }
 
+        /// Return the index a slice of potentially-invalid UTF-16 is normalized up to
+        pub fn is_normalized_utf16_up_to(&self, text: &[u16]) -> usize {
```

Comment:
When considering failure modes (cases which are reachable but only with weird input) I tend to favor the solution that keeps well-formed text running at full performance and does some well-defined but potentially GIGO behavior in ill-formed text.

Reply:
I've thought about this more:

First, the Web-exposed ICU4C-backed behavior of current `String.prototype.normalize` in both SpiderMonkey and V8 retains unpaired surrogates in the normalization process (even after the first point in the string that needs to change under normalization). We've previously decided that ICU4X operates on the Unicode Scalar Value / Rust `char` value space and, therefore, will perform replacement of unpaired surrogates with the REPLACEMENT CHARACTER.

I think we should not reopen that discussion and should take the marginal Web compat risk from switching from ICU4C normalization to ICU4X normalization behavior when it comes to unpaired surrogates.

A performance-neutral solution that allows `is_normalized_up_to`-based copy-on-write ("copy head code-unit-wise and use `normalize_to` on tail") to code-unit-wise match the outcome of using `normalize_to` on the whole input in a non-copy-on-write-manner can be implemented as follows:

1. Copypaste the UTF-16 and UTF-8 sinks under a new name in order not to disturb `is_normalized_utf16/8` semantics before 2.0. (Whether and when to disturb the semantics of those is a discussion outside this GitHub comment.)
2. On `utf16_iter::Utf16Chars` and `utf8_iter::Utf8Chars`, introduce (by copypaste; not sure if delegating the main `next` to the new method would be performance-neutral) a method `fn next_result(&mut self) -> Option<Result<char,Utf16Error>>` (or `Utf8Error`) that returns an error on not-well-formed input (naming is a bikeshed, but if there's precedent for naming, I'd like to hear about it). The error type should be a empty struct that carries no error detail so that `std::mem::size_of::<char>() == std::mem::size_of::<Result<char,Utf16Error>>()` holds.
3. In the sink copies from step 1, in `write_char`, change `iter.next() == Some(c)` to `iter.next_result() == Some(Ok(c))`. (AFAICT, this should be a performance-neutral change as long as the error type parameter of the `Result` is chosen so that the above `size_of` equality holds.)

We should settle the naming of the new method and the error case type parameter here so that I can release new versions of `utf16_iter` and `utf8_iter` without having to revise naming later.

(The above does not require changes to the normalization algorithms. The change is scoped to `write_char`, because the normalization algorithms themselves promise to output only well-formed UTF-16/8 via `write_slice`/`write_str`.)

(It seems more useful to pursue this than to bother with an `experimental` intermediate state.)

Reply:
("Performance-neutral" in the absence of unpaired well-formedness errors, that is.)

Reply:
> (by copypaste; not sure if delegating the main `next` to the new method would be performance-neutral) 

Godbolting the most obvious delegating implementation indeed shows and additional branch.

Reply:
Would things work with a utf16 iterator that returned `u32` instead of `char`, returning unpaired surrogates as integer values and paired surrogates as code points?

As you note, this is how the Web platform already works. See [WTF-8](https://simonsapin.github.io/wtf-8/)

Reply:
> I think we should not reopen that discussion and should take the marginal Web compat risk from switching from ICU4C normalization to ICU4X normalization behavior when it comes to unpaired surrogates.
> ...
> Copypaste the UTF-16 and UTF-8 sinks under a new name in order not to disturb is_normalized_utf16/8 semantics before 2.0. (Whether and when to disturb the semantics of those is a discussion outside this GitHub comment.)

A couple of thoughts here:

1. I'm more concerned about web compat risk than ICU4X compat risk right now, especially since 2.0 is coming up soon.
2. Although the behavior of unpaired surrogates is in principle out of scope of this PR, if we decide to go in this direction, I would rather have @CanadaHonk implement that solution instead of inventing some interim `Option<Result<char>>` solution so that we don't need to re-engineer this code again.

Reply:
> I'm more concerned about web compat risk than ICU4X compat risk right now

Are you suggesting that we overturn the previous decision that ICU4X won't output not-well-formed UTFs and in input will not distinguish different UTF errors from each other (i.e. all are treated as the REPLACEMENT CHARACTER)?

Is there evidence that the Web relies on ICU4C's unpaired surrogate round tripping in `String.prototype.normalize`?

It's somewhat unfortunate that to make new copy-on-write output consistent with pre-existing non-copy-on-write output would mean that assert_eq!(`text.len() == norm.is_normalized_utf16_up_to(text), norm.is_normalized_utf16(text)`) would not hold without adjusting the semantics of `is_normalized_utf16`.